### DISCUSSION TITLE
Interactive stdin

### DISCUSSION THREAD

**Discussion by AlphaNecron** [criyle/go-judge#73] [Q&A]
Is it even possible to pipe standard input to the isolated program? I'm trying to implement interactive mode for problems.

**Comment by AlphaNecron** [comment]
Btw, can we run x instances of judge on x cores? That would be great if we could customize the core the judge should run on.

**Reply by criyle** [reply]
The default parallelism equals to the number of cores. For a specific request, with cgroup enabled, you can use the `cpuSetLimit` to limit the cpu core to use. For example, `cpuSetLimit:“0,2”` restrict program to run only on cpu 0 and 2 while `cpuSetLimit:“0-3”` makes it utilizing cpu 0 to 3.

**Reply by AlphaNecron** [reply]
Oh, is it possible with seccomp?

**Reply by AlphaNecron** [reply]
For instance, I want to have 8 judges within a process, each bound to a specific core. Can it be done with seccomp?

**Reply by criyle** [reply]
Yes but you would need to build your own binary with `-tag seccomp` as it is disabled by default.

**Reply by AlphaNecron** [reply]
I'm implementing `go-sandbox` within my own project so I can directly use seccomp, so how to do it with seccomp?

**Reply by criyle** [reply]
You can pass the compiled seccomp BPF into https://github.com/criyle/go-judge/blob/master/env/linuxcontainer/environment_linux.go#L29 to apply seccomp for every execution.

**Reply by AlphaNecron** [reply]
Oh, so we're combining `cgroup` and `seccomp` in order to restrict CPU usage to a specific core?

**Reply by criyle** [reply]
In order to restrict CPU usage, you need to enable CPU share cgroup. To restrict CPU cores to use, you need to enable CPU set cgroup. To restrict syscall, you need seccomp.

**Reply by AlphaNecron** [reply]
I meant restricting allowed CPU cores. Anyways, thanks!

**Comment by criyle** [comment]
You may use the streaming protobuf interface to pipe stdin and stdout. Or you may create fifo pipe file and pass it as file with `pipe:true`. You may look at the executershell that demos an interactive shell in the isolated environment.

**Reply by AlphaNecron** [reply]
Thanks, gonna try it.

**Comment by AlphaNecron** [comment]
Btw, CPU restriction refers to logical cores or physical cores?

**Reply by criyle** [reply]
Logical cores

**Reply by AlphaNecron** [reply]
Oh, so theoretically we can run 16 judges on a 16 threads CPU?

**Reply by criyle** [reply]
Yes, the default parallelism equals to the number of logical cores and you can adjust it by command line argument.

**Reply by AlphaNecron** [reply]
Thanks.

**Reply by AlphaNecron** [reply]
Is it possible to get the output as a variable directly?

**Reply by criyle** [reply]
Yes, you can specify the `stdout` in the `copyOut `parameter.

**Reply by AlphaNecron** [reply]
Thanks, gonna try.

**Comment by AlphaNecron** [comment]
![image](https://github.com/criyle/go-judge/assets/57827456/89f88cc6-d249-4333-868a-20f6dbfc59cb)
cgroup throws this, any ideas?

**Reply by zhouyusd** [reply]
The executorserver need root privilege to create cgroup.

**Reply by AlphaNecron** [reply]
Oh, thanks!
